### PR TITLE
Rfc5914

### PR DESCRIPTION
**Pull request by carl-wallace**
- fixed visibility for several structs from earlier work
- added RFC5914 encoders/decoders
- address cargo doc warnings (mostly with tags in ASN.1 definitions in comments)
- added partial SignedData implementation (to allow parsing certs-only messages in support of work on certification path builder)

### CODE REVIEW

File: pkcs7/src/CryptographicMessageSyntax2004.rs [hunk 1]
```diff
@@ -0,0 +1,534 @@
+//! Selected structures from RFC5652
+
+use x509::{AttributeTypeAndValue, Certificate, Name, SubjectKeyIdentifier};
```
Comment (npmccallum) [inline pkcs7/src/CryptographicMessageSyntax2004.rs]: I suspect that we don't want `pkcs*` crates to depend on `x509`. We probably need to refactor some of the types in that crate into a shared crate. For example, `AttributeTypeAndValue` and `Name` are likely to be needed in a variety of places. The same with x509 extensions (not relevant to this commit). This probably requires some discussion.

File: x509/src/validity.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ use der::{Decodable, Error, Result, Sequence};
 ///     notAfter       Time  }
 /// ```
 /// [RFC 5280 Section 4.1.2.5]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Sequence)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Sequence, PartialOrd, Ord)]
```
Comment (npmccallum) [inline x509/src/validity.rs:14]: All of these derive changes should be a separate pull request.

**Review by npmccallum — CHANGES_REQUESTED** [review]
Hi @carl-wallace ! Thanks for your contribution!

I'm a somewhat new contributor to RustCrypto. But as I was asked to do a review, I'm going to do one. I'll aim to be somewhat picky and @tarcieri can tell me if I'm being too stringent.

1. This series of commits is written in a "stream of consciousness" style. Instead, each commit should be a small, self-contained logical change. This aids in reviewers understanding what you're trying to accomplish.

2. Commit messages should follow the best practices [outlined here](https://cbea.ms/git-commit/). As it stands now, your commit messages do not format well in GitHub and this makes them difficult to read.

3. You should set up your editor to run `cargo fmt` on every save. This ensures that each commit is properly formatted and you don't need to have a huge final commit that does all the formatting. This kind of code churn is bad for git merge algorithms.

4. A pull request should not contain merge commits. If you need to update your changes to account for the latest upstream changes, you should do a `git fetch` followed by a `git rebase`. Having commits be small, self-contained and logical (see comment 1) will help with this process.

5. Once you have reformatted commits to be small, self-contained, and logical, then commits that are somewhat trivial should be filed as separate pull requests. These can be merged much quicker and reduce the amount of code you need to rebase. For example, deriving `PartialOrd` and/or `Ord` on types is a great candidate for a small, self-contained, logical commit that is eligible for immediate merge. The same is true for documentation warning fixes.

6. In numerous of your commits you make changes that aren't self-evident and are not explained. For example, when a type derives `Sequence` you remove the derive and choose to implement the trait manually. Why? This isn't explained. Deriving traits should be the default unless there is a compelling reason to do otherwise. If these were small, self-contained, logical commits you could explain the reason for these changes in the commit message body.

File: x509/src/certificate.rs [hunk 3]
```diff
@@ -202,6 +202,36 @@ pub struct Certificate<'a> {
...
+        F: FnOnce(&[&dyn der::Encodable]) -> ::der::Result<T>,
+    {
+        f(&[
+            &self.tbs_certificate,
+            &self.signature_algorithm,
+            &self.signature,
+        ])
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/certificate.rs:237]: That is possibly one that could revert to the macro, and we should try. Generally, places where I did not use the macros it's because the macros did not achieve the desired end. For some, I updated the macros, for some I did not. The macros could be revisited in most of those places (but this may be one that ought revert).

File: x509/src/pkix_extensions.rs [hunk 4]
```diff
@@ -328,30 +328,32 @@ pub struct PolicyMapping {
 pub struct NameConstraints<'a> {
     /// permittedSubtrees       [0]     GeneralSubtrees OPTIONAL,
     //#[asn1(context_specific = "0", optional = "true", tag_mode = "IMPLICIT")]
-    permitted_subtrees: Option<GeneralSubtrees<'a>>,
+    pub permitted_subtrees: Option<GeneralSubtrees<'a>>,
 
     /// excludedSubtrees        [1]     GeneralSubtrees OPTIONAL }
     //#[asn1(context_specific = "1", optional = "true", tag_mode = "IMPLICIT")]
-    excluded_subtrees: Option<GeneralSubtrees<'a>>,
+    pub excluded_subtrees: Option<GeneralSubtrees<'a>>,
```
[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/pkix_extensions.rs:366]: That's true, but they were arrived at later as other pieces were moving. I should have pulled those out though separate from the rest. Forewarning, the builder and validator are chock full of stream of consciousness:-) Point understood though.

File: x509/src/certificate.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply x509/src/certificate.rs:237]: I would propose that anywhere the macros don't work, an issue should be filed for the missing or broken functionality and a reference to the issue should be place in a comment next to the manual implementation.

File: x509/src/pkix_extensions.rs [hunk 5]
```diff
@@ -328,30 +328,32 @@ pub struct PolicyMapping {
...
+            ::der::asn1::ContextSpecific::decode_implicit(decoder, ::der::TagNumber::N0)?
+                .map(|cs| cs.value);
+        let excluded_subtrees =
+            ::der::asn1::ContextSpecific::decode_implicit(decoder, ::der::TagNumber::N1)?
+                .map(|cs| cs.value);
+        Ok(Self {
+            permitted_subtrees,
+            excluded_subtrees,
         })
+        // })
```
[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/pkix_extensions.rs:387]: Because it didn't work. We could revert and compare cargo expand to what I manually did. There have been some changes to derive since these edits were made so it is possible these manual edits are not needed.

File: x509/src/pkix_extensions.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (npmccallum) [reply x509/src/pkix_extensions.rs:366]: Stream of consciousness is perfectly fine for your local development branch. But once the tests are passing and you think it is ready for merge, you should generally do a `git reset upstream/master` followed by a `git add -p` to selectively create logical commits.

[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/pkix_extensions.rs:366]: The next three have no precursors, so perhaps less of an issue. Those will be born anew once repotted in the formats repo so stream of consciousness will be absent for those (I had forgotten that intermediate step on that looming PR).

File: x509/src/trust_anchor_format.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        number: TAC_TA_INFO_TAG,
+    };
+}
+
+// Not supporting these structures
+// TrustAnchorList ::= SEQUENCE SIZE (1..MAX) OF TrustAnchorChoice
+//
+// id-ct-trustAnchorList      OBJECT IDENTIFIER ::= { iso(1)
+//     member-body(2) us(840) rsadsi(113549) pkcs(1) pkcs9(9)
+//     id-smime(16) id-ct(1) 34 }
```
[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/trust_anchor_format.rs:340]: 1. I can cancel the PR and chunk it up. Part of the problem is having enough to test things is tricky with these. But certainly these could have been smaller.
2. The name is derived from the spec. If anything, we probably ought try to keep things from ASN.1 modules in files per module and use module name (but that ship has sailed).
3. I agree re: macros. However, there are spots where the macros don't quite work. Getting the macros to the point where one could just use them and them alone is a good goal, but it wasn't where I could spend all of my time. That there are manual indicators for places where macro work could be helpful is maybe a good thing.

File: x509/src/certificate.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (carl-wallace) [reply x509/src/certificate.rs:237]: I can open issues for each manual deviation. That's not a problem.

**Comment by carl-wallace** [comment]
Closing, will resubmit in smaller chunks.

**Comment by tarcieri** [comment]
Just one note on commit history and comments:

We use GitHub's "Squash and Merge" feature so that each commit maps 1:1 with a PR.

Because of that, it doesn't really matter how messy the branch history is, or if there are superfluous comments. It will all get squashed down to a single commit, and any superfluous comments edited away into a single commit message.
